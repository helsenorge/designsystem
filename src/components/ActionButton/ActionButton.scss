// @import '../../styling/palette';
// @import '../../styling/components/button';

// primary
$action-button-primary-color-font: $surgical-500;
$action-button-primary-color-background: $bone;
$action-button-primary-color-border: $surgical-500;
// primary:hover
$action-button-primary-hover-color-background: $surgical-100;
$action-button-primary-hover-color-box-shadow: $surgical-400;
$action-button-primary-hover-color-border: $surgical-100;
// primary:disabled
$action-button-primary-disabled-color-background: $ash;
$action-button-primary-disabled-color-font: $waiting-room;

// secondary
$action-button-secondary-color-font: $surgical-500;
$action-button-secondary-color-background: transparent;
$action-button-secondary-color-border: $surgical-500;
// secondary:hover
$action-button-secondary-hover-color-border: $surgical-100;
$action-button-secondary-hover-color-box-shadow: $surgical-400;
$action-button-secondary-hover-color-background: $surgical-100;
// secondary:disabled
$action-button-secondary-disabled-color-border: $waiting-room;
$action-button-secondary-disabled-color-font: $waiting-room;

// tertiary
$action-button-tertiary-color-font: $surgical-500;
$action-button-tertiary-color-background: transparent;
// tertiary:hover
$action-button-tertiary-hover-color-border: $surgical-100;
$action-button-tertiary-hover-color-box-shadow: $surgical-400;
$action-button-tertiary-hover-color-background: $surgical-100;
// tertiary:disable
$action-button-tertiary-disabled-color-border: $waiting-room;
$action-button-tertiary-disabled-color-font: $waiting-room;

.is-action {
  background-color: $action-button-primary-color-background;
  border: 0.125rem solid $action-button-primary-color-border;
  color: $action-button-primary-color-font;
  @include button-variant-borders(
    $action-button-primary-hover-color-border,
    $action-button-primary-hover-color-box-shadow
  );
  &:disabled,
  &.is-disabled {
    color: $action-button-primary-disabled-color-font;
    background-color: $action-button-primary-disabled-color-background;
  }
  &:hover,
  &.is-hovered {
    background-color: $action-button-primary-hover-color-background;
    border-color: $action-button-primary-hover-color-border;
  }
  &:active,
  &.is-active {
    background-color: $action-button-primary-hover-color-background;
    border-color: $action-button-primary-hover-color-border;
  }
  &:focus,
  &.is-focused {
    background-color: $action-button-primary-hover-color-background;
    border-color: $action-button-primary-hover-color-border;
  }
  &.is-secondary {
    @include button-variant-borders(
      $action-button-secondary-hover-color-border,
      $action-button-secondary-hover-color-box-shadow
    );
    background-color: $action-button-secondary-color-background;
    border: 0.125rem solid transparent;
    color: $action-button-secondary-color-font;
    &:disabled,
    &.is-disabled {
      color: $action-button-secondary-disabled-color-font;
      border-color: $action-button-secondary-disabled-color-border;
    }
    &:hover,
    &.is-hovered {
      background-color: $action-button-secondary-hover-color-background;
      border: 0.125rem solid $action-button-secondary-hover-color-border;
    }
    &:active,
    &.is-active {
      background-color: $action-button-secondary-hover-color-background;
      border: 0.125rem solid $action-button-secondary-hover-color-border;
    }
    &:focus,
    &.is-focused {
      background-color: $action-button-secondary-hover-color-background;
      border: 0.125rem solid $action-button-secondary-hover-color-border;
    }
  }
  &.is-tertiary {
    border: 0;
    background-color: $action-button-tertiary-color-background;
    color: $action-button-tertiary-color-font;
    &:disabled,
    &.is-disabled {
      border: 0;
    }
    &:hover,
    &.is-hovered &:active,
    &.is-active,
    &:focus,
    &.is-focused {
      &:before {
        border: 0;
      }
      &:after {
        border: 0;
      }
    }
  }
}
