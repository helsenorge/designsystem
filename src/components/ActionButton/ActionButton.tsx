import React from 'react';
import cn from 'classnames';

// import './ActionButton.scss';

// TODO: Make this inherit from a set of base variant types in constants.
export type ActionButtonVariant = 'seconary' | 'tertiary' | string | undefined;

// TODO: Make this inherit attributes form both 'a'-tags and 'button'-tags.
interface ActionButtonProps
  extends React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  children: React.ReactNode;
  variant?: ActionButtonVariant;
  isLoading?: boolean;
  isDanger?: boolean;
}

// TODO: Move most of the logic out in a generic Button-component that the others inherit.
const ActionButton = React.forwardRef((props: ActionButtonProps, ref: any) => {
  const {children, variant, isLoading = false, isDanger = false, ...rest} = props;
  const classes = cn(
    'button is-action',
    {[`is-${variant}`]: variant === 'secondary' || variant === 'tertiary'},
    {['is-loading']: isLoading},
    {['is-danger']: isDanger},
  );
  return (
    // TODO: Add a 'as' prop so that the button can be either an 'a'-tag or 'button'-tag.
    <button className={classes} ref={ref} {...rest}>
      <span>{children}</span>
    </button>
  );
});

ActionButton.defaultProps = {
  type: 'button',
};

export {ActionButton};
