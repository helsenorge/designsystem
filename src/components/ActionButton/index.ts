export {ActionButton, ActionButtonVariants} from './ActionButton';
