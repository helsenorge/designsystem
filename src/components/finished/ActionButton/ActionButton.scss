@import '../../../styling/colors';

.action-button {
  cursor: pointer;
  font-size: 1.125rem;
  letter-spacing: 0.03125em;
  padding: 0.65rem 1.5rem 0.55rem;
  border-radius: 1.5rem;
  appearance: none;
  &--primary {
    background-color: $primary-color;
    color: white;
    border: 3px solid transparent;
    &:focus:not([disabled]),
    &:hover:not([disabled]) {
      background-color: $primary-color-hover;
      box-shadow: 0 0 0 4px $outline;
      outline: none;
    }
    &:disabled {
      background-color: $disabled;
      color: $disabled-text;
    }
  }
  &--secondary {
    background-color: transparent;
    border: 3px solid $primary-color;
    color: $primary-color;
    &:focus:not([disabled]),
    &:hover:not([disabled]) {
      background-color: transparent;
      border: 3px solid $primary-color-hover;
      box-shadow: 0 0 0 4px $outline;
      color: $primary-color-hover;
    }
    &:disabled {
      color: $disabled;
      border: 3px solid $disabled;
    }
  }
  &--tertiary {
    color: primary-color;
    border: none;
    border: 3px solid transparent;
    &:focus:not([disabled]),
    &:hover:not([disabled]) {
      box-shadow: 0 0 0 4px $outline;
      outline: none;
    }
    &:disabled {
      color: $disabled;
    }
  }
}
