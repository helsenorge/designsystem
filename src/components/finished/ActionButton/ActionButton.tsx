import cn from 'classnames';
import React from 'react';
import {ButtonVariants} from '../../../constants';
import styles from './ActionButton.scss';

interface ActionButtonProps extends React.HTMLProps<HTMLButtonElement> {
  children: React.ReactNode;
  onClick?: () => void;
  variant?: ActionButtonVariants;
}

function ActionButton({children, variant = ButtonVariants.Primary, ...restProps}: ActionButtonProps): JSX.Element {
  const classes = cn(
    styles['action-button'],
    {[styles['action-button--primary']]: variant === ButtonVariants.Primary},
    {[styles['action-button--secondary']]: variant === ButtonVariants.Secondary},
    {[styles['action-button--tertiary']]: variant === ButtonVariants.Tertiary},
  );

  return (
    <button className={classes} {...restProps}>
      {children}
    </button>
  );
}

ActionButton.defaultProps = {
  type: 'button',
};

export type ActionButtonVariants = (typeof ButtonVariants)[keyof typeof ButtonVariants];
export default ActionButton;
