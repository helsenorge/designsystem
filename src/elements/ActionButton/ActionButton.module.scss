@import '../../styling/colors';

.button {
    display: block;
    cursor: pointer;
    background-color: $primary-color;
    color: white;
    white-space: normal;
    text-overflow: ellipsis;
    font-size: 1.125rem;
    font-weight: 400;
    letter-spacing: 0.03125em; /* 16 * 0.0625 = 1px */
    padding: 0.65rem 1.5rem 0.55rem;
    text-align: center;
    border: 3px solid transparent;
    border-radius: 1.5rem;
    appearance: none;
    &:focus,
    &:hover {
        background-color: $primary-color-hover;
        box-shadow: 0 0 0 4px $outline;
    }
    &.button--secondary {
        background-color: transparent;
        border: 3px solid $primary-color;
        color: $primary-color;
        &:focus,
        &:hover {
            border: 3px solid $primary-color-hover;
            color: $primary-color-hover;
        }
    }
}