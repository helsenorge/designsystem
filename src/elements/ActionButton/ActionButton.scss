@import '../../styling/colors';

.action-button {
  position: relative;
  display: inline-flex;
  cursor: pointer;
  white-space: normal;
  font-size: 1.125rem;
  font-weight: 400;
  letter-spacing: 0.03125em; /* 16 * 0.0625 = 1px */
  padding: 0.65rem 1.5rem 0.55rem;
  text-align: center;
  align-items: center;
  justify-content: center;
  border-radius: 1.5rem;
  appearance: none;
  &--primary {
    background-color: $primary-color;
    color: white;
    border: 3px solid transparent;
    &:focus:not([disabled]),
    &:hover:not([disabled]) {
      background-color: $primary-color-hover;
      box-shadow: 0 0 0 4px $outline;
      outline: none;
    }
    &:disabled {
      background-color: $disabled;
      color: $disabled-text;
    }
  }
  &--secondary {
    background-color: transparent;
    border: 3px solid $primary-color;
    color: $primary-color;
    &:focus:not([disabled]),
    &:hover:not([disabled]) {
      background-color: transparent;
      border: 3px solid $primary-color-hover;
      box-shadow: 0 0 0 4px $outline;
      color: $primary-color-hover;
    }
    &:disabled {
      color: $disabled;
      border: 3px solid $disabled;
    }
  }
  &--tertiary {
    color: primary-color;
    border: none;
    border: 3px solid transparent;
    &:focus:not([disabled]),
    &:hover:not([disabled]) {
      box-shadow: 0 0 0 4px $outline;
      outline: none;
    }
    &:disabled {
      color: $disabled;
    }
  }
}
