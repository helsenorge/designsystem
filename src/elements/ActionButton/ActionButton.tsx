import cn from 'classnames';
import React from 'react';
import styles from './ActionButton.module.scss';

type ActionButtonVariant = 'primary' | 'secondary';

interface ActionButtonProps extends React.HTMLProps<HTMLButtonElement> {
  children: React.ReactNode;
  onClick?: () => void;
  variant?: ActionButtonVariant;
}

function ActionButton(props: ActionButtonProps) {
  const classes = cn(styles.button, {[styles['button--secondary']]: props.variant === 'secondary'});
  return (
    <button className={classes} {...props}>
      {props.children}
    </button>
  );
}

ActionButton.defaultProps = {
  type: 'button',
  variant: 'primary',
};

export {ActionButton};
