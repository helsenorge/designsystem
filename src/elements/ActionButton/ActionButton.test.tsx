import React from 'react';
import {render} from 'react-testing-library';
import ActionButton from '.';

test('default variant renders correctly', () => {
  const {container} = render(<ActionButton>Hello World</ActionButton>);
  expect(container).toMatchSnapshot();
});
