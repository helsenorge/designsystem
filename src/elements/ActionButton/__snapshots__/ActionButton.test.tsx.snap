// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <button
    class="button"
    type="button"
    variant="primary"
  >
    Hello World
  </button>
</div>
`;
